### ISSUE TITLE
Implement openface through tensorflow to get closer to 1800 batch size

### ISSUE THREAD

**Issue by myme5261314** [cmusatyalab/openface#129]
I've written some code using tensorflow and attempt to make triplets generating from 45 person * 40 images. I've roughly done with the code. But the experiments shows the result as wrong. I've been working on this issue for about 2-3 weeks. Maybe there are some guys that are interested to work together with me to work out the tensorflow `facenet`? Thanks to @davidsandberg for his code as a base.

----

> 
> On Fri, Apr 29, 2016, 3:44 AM Peng Liu <myme5261314@gmail.com> wrote:
> Dear Bamos,
>     I'm now working on tensorflow implementation of FaceNet, I've read through your openface code and https://github.com/davidsandberg/facenet code. I don't know if you are familiar with tensorflow. Honestly, I'm not very good at Torch.
>     It seems that tensorflow could help use 45*40 images per batch and all possible invalid triplets without reduce. The reason why tensorflow could and Torch couldn't is that tensorflow has API to compute, retrieve gradients, and apply gradients separately. And this makes the possible space-time trade-off. Let me explain abstractly.
>     1. Use a batch with 45 people, 40 images per people, this will be 1800 images. Use a mini-batch size say 90.
>     2. Forward batch via 20 (45*40/90) mini-batches to get all the 1800 embeddings.
>     3. Using 1800 embeddings to generate all possible triplets cost.
>     4. Sample the triplets you want to use.
>     5. Use sampled triplets to get 1800 gradient coefficients.
>     6. Use 90 images and 90 gradient coefficients a time to calculate gradients to all trainable variables. repeat 20 times.
>     7. accumulate these gradients, and then apply them via optimizer.
> 
>     I\'ve implemented all these 7 steps code in my fork of "facenet", but the code seems to have some problems or even my theory is wrong? I really have no idea how to correct these code since I have tried that for about 2 weeks.
>     So, do you have any idea about this?
>     Please refer to some "facenet" related info to https://github.com/davidsandberg/facenet/issues/14
> 
> Yours Sincerely,
> Peng Liu
>

**Comment by myme5261314** [comment]
Is there anyone who's interested in achieving the trade-off to use current limited GPU memory while using 1800 images to generate triplets?
[my fork](https://github.com/myme5261314/facenet)

**Comment by melgor** [comment]
Hi,
unfortunately, I am not able to help you (I do not know the TensorFlow at all). But I have couple of questions:
1. As I understand, you want to connect 2 training ideas from OpenFace: initial training code (batch size 1800, then selecting triplets and again forward pass for each triplets) with second version (generating triplets from embedding). I think it is nice idea and this could be done in Torch.
2. You aim to get batchsize=1800. But what happened when you have batch size = 90? I mean when you have only one forward-backward pass. Does it works then? You must find the place, where there is an issue (train code, triplets generation, merging gradient, etc)
3. The last thing is about selecting triplets. Your current version is biased on initial training code. I think it was the main issue of very low quality of models (~77% on LFW). I was changes to works with no random negative examples. Why? Because each random negative example produce loss =0, so gradient is equal 0 too. If you have a lot of random negatives, your results will be really low. So you should only choose triplets without random negatives. Second thing is formula
```python
if pos_dist<neg_dist and neg_dist<sel_neg_dist and np.abs(pos_dist-neg_dist)<FLAGS.alpha:
```
 is not working well at practice. It is connected with 'random negative' examples. We changed it to VGG idea, so choosing any negative examples which is at margin (so neg_dist can be < than pos_dist). At it works much better.

To sum up, I advise you:
1. See it everything works with signle Forward-Backward pass
2. Change idea of choosing triplets to VGG style and only choosing triplets without random negatives

**Comment by myme5261314** [comment]
Thanks for your reply, @melgor . Actually, I haven't made it clear that. My `facenet` repo is forked from @davidsandberg . I have add many new things based on that to my branch [optimize](https://github.com/myme5261314/facenet/tree/optimize) while the master branch for now is still of @davidsandberg efforts.
Maybe when you have time, please check the optimize branch.
As I stated in the email, I've implemented the following functionalities.

1. Using 1800 embeddings to generate all possible triplets cost.
2. Sample the triplets you want to use.
3. Use sampled triplets to get 1800 gradient coefficients for 1800 samples per batch

So, I can use **any** `sample` strategy to downsample the all 45*40*45*40*40 triplets (just all the combinations are contained no matter valid or not.)

After I get 1800 gradient coefficients, I can use 20 mini-batch (90) to get 20 gradients (tensorflow has the API to get gradient and also apply gradient rather than just backward of `Torch`), then accumulate these 20 groups of gradients, and then apply the result gradient once. Then one batch (1800) is over.

So, I've confirmed the triplets generating and downsampling and coefficients calculation are all correct since I've wrote some test code to assert with the simulated simple situation.

I've digged into the very low level code of both torch, tensorflow, openface and `facenet`. It's so pathetic that I haven't found the true reason.

At first I suppose it was the `batch normalization` implemented wrong in the `facenet` makes the result wrong. Then I fixed that. But result is still not good. Then I found removing `batch normalization` layers all the way would pass the gradient checking code I wrote while failed with those `batch normalization layer`, it's really easy to understand. `batch normalization` makes the irrelevant before samples of batches relevant. But I don't think removing all the `batch normalization` is a good idea. I mean, `openface` is the direct evidence that using `gradient coefficients` to calculate the gradients is equivalent to expanding all the triplets loss.

At least, I think maybe implementing the current functionalities to tensorflow version `facenet` is a base goal. I mean, just like current `openface`, use 15*20 images per batch, use the same model architecture the openface used, use the same dataset, and then use the same hyper-parameters, and finally achieve the same accuracy on LFW as openface did. Then, I can move on to really "optimize" my `optmize` branch.

I run the experiment of `openface` today, with default hyper-parameters, it will achieve around 70% accuracy on LFW after one epoch, but my fork stucks into around 50%, I really don't have any new idea to debug my code now.

Maybe I need to give up my new content away and re-implement via tensorflow from zero to get what openface have now achieved.

**Comment by myme5261314** [comment]
I have fixed my issue by myself. So close this issue.

**Comment by melgor** [comment]
So what was the bug? I go through the code I can not find anything (at my level of understanding of TF).
Do you any similar/better results than OpenFace?

**Comment by myme5261314** [comment]
There are **many** minor errors or mismatch in the buggy code.
Some exists when I forked the repo from @davidsandberg, such as the `batch_norm` code.
You see, @davidsandberg starts the original `facenet` repo according to the early version of `openface`, `openface` is keeping improving and changing. But he seems have not enough time to spend on it when he achieved 92% with a very different and weird approach compared to openface in my opinion. And finally, I understand him. It's the difference between two libraries' API (`tensorflow` and `torch`) that makes him not able to simply re-implement `openface` via `tensorflow` as with `torch`.

So, I had two versions of "FaceNet", and they all get around 92% on LFW at the beginning. But I need to re-implement and also extend the "FaceNet" to be more closer to Google's paper. The bug seems all occur by kind of modifying `facenet` one place according to `openface`, but forget to modify another place.

Speaking about result. It's not strictly comparable, since I use both CASIA and Facescrub, but using `openface` dlib scripts to squeeze out the people whose images is less than 40. Because I attempt to use 45\*40 images. So, I try to use the same or similar experiment parameters as `openface` and get around 92% on LFW. This result can kind of validate my implementation now, right?

I'm now running several experiments now. One is using 1800 images per batch and downsample by random to choose learning triplets. The other is using 300 images per batch and use all possible triplets which contribute to triplets loss.

The experiments are still running. Since my `facenet` code is kind of 2-3 times slower than `openface`, the experiments might take some time.

Just for advice, if `openface` want to use 45*40, it requires API of torch to implement calculating and retrieving gradients, I'm not familiar with `torch` so I don't know if it's capable.

**Comment by davidsandberg** [comment]
Hi Peng!
I'm glad to hear that you have had some progress on this!
I have fixed the bug in the batch normalization. Thanks for pointing this out!
From the first experiments it looks like the performance is still ok.
I saw that you have added Local Response Normalization. Did you compare the performance with/without LRN to see if you got any performance gain from it? Also, did you see any improvements from AdaDelta optimizer compared to AdaGrad?
Have you thought about getting a bigger training set than the FS+Casia?
I made a small python program to download the VGG Face dataset:
https://github.com/davidsandberg/facenet/blob/master/facenet/src/download_vgg_face_dataset.py
But the download takes time and it looks like it contains quite a lot of duplicates (which doesn't degrade performance according to the VGG Face paper). But before I start using this dataset I have a bunch of performance comparisons to do, like L2 pooling, AdaDelta, Dropout, L2 penalty, LRN etc. 
BR
David

**Comment by myme5261314** [comment]
Hi David! I think the most important thing that you need to do is to upgrade your triplets generating strategy. I mean you are using the strategy that is same with early version of openface, i.e. use only only image of each person as a anchor. This strategy is not reasonable, right? Even though, you can still get accuracy around 92%.
Talking about LRN and AdaDelta, I'm just following the existing experience from `openface`, i.e. they are using this. So, I haven't gone through these comparison.
As I answered to @melgor , I'm running two experiments simultaneously now. And because of the slow performance of tensorflow compared to torch, it will still take a few days to finish these two experiments even though they have been running for 3-4 days so far.
It's really depressed that `tensorflow` is such slower than `torch`, and I don't have the ability to fix that. More than 80% running time is consuming by the tensorflow API, so no matter how much optimization I use on other codes, it's not worth it.
Please show us your conclude when you have some progress on those performance comparisons you mentioned.
Will discuss with you two when I have some conclusion of my current two experiments.

**Comment by myme5261314** [comment]
@davidsandberg , by the way, I've finished VGG dataset downloading before April, 2016. From what I get, there are many dead links and blank images or images with multiple people in it. Yes, very dirty.

**Comment by davidsandberg** [comment]
In the facenet paper they talk about using all the anchor-positive pairs, i.e. 
"Instead of picking the hardest positive, we use all anchor-positive pairs in a mini-batch while still selecting the hard negatives". And that is what we do in the tensorflow implementation (in select_validation_triplets) as well.
And due to shuffling in the sample_people function, the anchor images will be different each time triplets are selected.
I don't see that this part should be too far from what they did in the facenet paper. And I would think that the negative selection would be even more interesting to look into.

With respect to training time I saw a ~100% performance increase when I went from TF r0.7 to TF r0.8. I haven't used Torch for a while so i don't know if it has been a similar performance increase lately. But I would be supprised if that is the case, since as you say, most of the number crunching is done using CuDNN.

I will get back to you when I have some more results to show.

**Comment by RomainSabathe** [comment]
Hi everyone, thanks for the very interesting discussion. I have been interested lately in using the triplet loss for a slightly different problem, but still related.
Anyways, I had a strategy  in mind to find valid/good triplets but it seems so simple that I must be forgetting something.
Say the batch size is 64 and is composed of 8 classes * 8 pictures (8 persons with 8 photos for each person). I compute the embedding for each picture and  generate *all* possible embedding triplets (a,p,n) which makes 8 * 8 * 7 * 7 * 8 =  25 088 possibilities. The associated "metric" is computed for each triplet: max(0, alpha - (a - n)**2 + (a - p)**2). I count the number of non-null metrics (which therefore corresponds to the number of semi-hard triplets), say N. I finally take the loss to be sum(all metrics) / N.
Since we don't take the mean of all those metrics strictly speaking, it prevents the error to artificially drop due to the decreasing number of semi-hard.

To what extent do you think this could be a valid approach? Or, say, an invalid approach? :)

**Comment by myme5261314** [comment]
you should really notice that the downsample strategy from VGG paper。

发自我的 iPhone
